### PR TITLE
Calculating gradients for partial graph

### PR DESCRIPTION
**Pull request by emailweixu**
Added backward.calc_gradient to backpropagate gradient from given targets to inputs.

### CODE REVIEW

File: python/paddle/v2/fluid/backward.py [hunk 1]
```diff
@@ -340,3 +380,147 @@ def append_backward(loss, parameter_list=None, no_grad_set=None):
...
+    if inputs:
+        for op in op_path:
+            for name in op.desc.input_arg_names():
+                if name not in input_names:
+                    no_grad_set.add(name)
+
+    return op_path
+
+
+def calc_gradient(targets, inputs, target_gradients=None, no_grad_set=None):
```
Comment (JiayiFeng) [inline python/paddle/v2/fluid/backward.py]: I have a few questions about the `calc_gradient`:
1. It seems that this API appends partial backward to calculate gradients of `targets` to `inputs`. Can this API be used by users directly? And when will users use it?
2. Here are some special handlings(like `_rename_grad_`) to make sure this API can be invoked for multiple times. However, in each invoking the `program` itself will be changed(some backward ops are appended). So when we invoke it second this, the backward part generated in the first time is still in it and are regarded as forwarding part. Is this reasonable?

Comment (emailweixu) [reply python/paddle/v2/fluid/backward.py]: 1. Yes. Some algorithm have its own way of calculating gradients. One example is [Learning to learn by gradient descent by gradient descent](https://arxiv.org/abs/1606.04474).
2. Yes. This happens in cases where gradients from different parts needs to be combined in some way to calculate the final result.

Comment (JiayiFeng) [reply python/paddle/v2/fluid/backward.py]: Got it!
